libraries/AP_LTM_Telem/AP_LTM_Telem.cpp

```diff
@@ -0,0 +1,234 @@
+        AP_SERIALMANAGER_LTM_BUFSIZE_TX);
+        hal.scheduler->register_io_process(
+                FUNCTOR_BIND_MEMBER(&AP_LTM_Telem::tick, void));
+    }
+}
+
+void AP_LTM_Telem::send_LTM(uint8_t lt_packet[], uint8_t lt_packet_size)
+{
+    // check space before write
+    if (_port->txspace() < lt_packet_size) {
+        return;
+    }
+    // calculate checksum
+    uint8_t lt_crc = 0x00;
+    for (uint8_t i = 3; i < lt_packet_size - 1; i++) {
+        lt_crc ^= lt_packet[i];
+    }
+    lt_packet[lt_packet_size - 1] = lt_crc;
+    _port->write(lt_packet, lt_packet_size);
+}
+
+// GPS frame
+void AP_LTM_Telem::send_Gframe(void)
+{
+    int32_t lat;                        // latitude
+    int32_t lon;                        // longtitude
+    uint8_t gndspeed;                   // gps ground speed
+    int32_t alt;                        // altitude
+    uint8_t sats_visible;               // number of visible gps sats
+    uint8_t fix_type;                   // gps fix type
+    Location loc;
+    float alt_ahrs;
+    const AP_GPS &gps = AP::gps();
+    const AP_AHRS &ahrs = AP::ahrs();
+
+    ahrs.get_relative_position_D_home(alt_ahrs);
+
+    if (ahrs.get_position(loc)) {
+        lat = loc.lat;
+        lon = loc.lng;
+        gndspeed = (uint8_t) roundf(gps.ground_speed());
+        alt = (int32_t) roundf(alt_ahrs);
+        sats_visible = (uint8_t) AP::gps().num_sats();
+        fix_type = (uint8_t) AP::gps().status();
+    } else {
+        lat = 0;
+        lon = 0;
+        gndspeed = 0;
+        alt = (int32_t) roundf(alt_ahrs);
+        sats_visible = (uint8_t) AP::gps().num_sats();
+        fix_type = (uint8_t) AP::gps().status();
+    }
+
+    uint8_t lt_buff[LTM_GFRAME_SIZE];
+    // protocol: START(2 bytes)FRAMEID(1byte)LAT(cm,4 bytes)LON(cm,4bytes)SPEED(m/s,1bytes)ALT(cm,4bytes)SATS(6bits)FIX(2bits)CRC(xor,1byte)
+    // START
+    lt_buff[0] = 0x24; //$
+    lt_buff[1] = 0x54; //T
+    // FRAMEID
+    lt_buff[2] = 0x47; // G ( gps frame at 5hz )
+    // PAYLOAD
+    lt_buff[3] = (lat >> 8 * 0) & 0xFF;
+    lt_buff[4] = (lat >> 8 * 1) & 0xFF;
+    lt_buff[5] = (lat >> 8 * 2) & 0xFF;
+    lt_buff[6] = (lat >> 8 * 3) & 0xFF;
+    lt_buff[7] = (lon >> 8 * 0) & 0xFF;
+    lt_buff[8] = (lon >> 8 * 1) & 0xFF;
+    lt_buff[9] = (lon >> 8 * 2) & 0xFF;
+    lt_buff[10] = (lon >> 8 * 3) & 0xFF;
+    lt_buff[11] = (gndspeed >> 8 * 0) & 0xFF;
+    lt_buff[12] = (alt >> 8 * 0) & 0xFF;
+    lt_buff[13] = (alt >> 8 * 1) & 0xFF;
+    lt_buff[14] = (alt >> 8 * 2) & 0xFF;
+    lt_buff[15] = (alt >> 8 * 3) & 0xFF;
+    lt_buff[16] = ((sats_visible << 2) & 0xFF)
+            | ((fix_type > 3 ? 3 : fix_type) & 0b00000011); // last 6 bits: sats number, first 2:fix type (0, 1, 2, 3)
+    send_LTM(lt_buff, LTM_GFRAME_SIZE);
+    _ltm_scheduler++;
+}
+
+// Sensors frame
+void AP_LTM_Telem::send_Sframe(void)
+{
+    uint16_t volt;                      // battery voltage (mV)
+    uint16_t amp;                       // actual current in mAh
```

Comment:
```suggestion
```

Defining these as laste as possible (and using const where possible) is best AP practice.

Reply:
(just one example, obviously)

Reply:
More acknowledgement from Ghettostation is added to first PR's comment, especially debug output log: https://github.com/ArduPilot/ardupilot/files/3567121/03-station-serial-debug.txt